README.md
generic-session
=========

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Coveralls][coveralls-image]][coveralls-url]
[![David deps][david-image]][david-url]
[![node version][node-image]][node-url]
[![npm download][download-image]][download-url]
[![Gittip][gittip-image]][gittip-url]

[npm-image]: https://img.shields.io/npm/v/koa-generic-session.svg?style=flat-square
[npm-url]: https://npmjs.org/package/koa-generic-session
[travis-image]: https://img.shields.io/travis/koajs/generic-session.svg?style=flat-square
[travis-url]: https://travis-ci.org/koajs/generic-session
[coveralls-image]: https://img.shields.io/coveralls/koajs/generic-session.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/koajs/generic-session?branch=master
[david-image]: https://img.shields.io/david/koajs/generic-session.svg?style=flat-square
[david-url]: https://david-dm.org/koajs/generic-session
[node-image]: https://img.shields.io/badge/node.js-%3E=_6.0.0-red.svg?style=flat-square
[node-url]: http://nodejs.org/download/
[download-image]: https://img.shields.io/npm/dm/koa-generic-session.svg?style=flat-square
[download-url]: https://npmjs.org/package/koa-generic-session
[gittip-image]: https://img.shields.io/gittip/dead-horse.svg?style=flat-square
[gittip-url]: https://www.gittip.com/dead-horse/

Generic session middleware for koa, easy use with custom stores such as [redis](https://github.com/koajs/koa-redis) or [mongo](https://github.com/freakycue/koa-generic-session-mongo), supports defer session getter.

This middleware will only set a cookie when a session is manually set. Each time the session is modified (and only when the session is modified), it will reset the cookie and session.

You can use the rolling sessions that will reset the cookie and session for every request which touch the session.  Save behavior can be overridden per request.

**For async/await and Node v6.9.0+ support use v2.x of this package, for older use v1.x**

## Usage

### Example

```js

var session = require('koa-generic-session');
var redisStore = require('koa-redis');
var koa = require('koa');

var app = new koa(); // for koa v1 use `var app = koa();`
app.keys = ['keys', 'keykeys'];
app.use(session({
  store: redisStore()
}));

app.use(function *() {
  switch (this.path) {
  case '/get':
    get.call(this);
    break;
  case '/remove':
    remove.call(this);
    break;
  case '/regenerate':
    yield regenerate.call(this);
    break;
  }
});

function get() {
  var session = this.session;
  session.count = session.count || 0;
  session.count++;
  this.body = session.count;
}

function remove() {
  this.session = null;
  this.body = 0;
}

function *regenerate() {
  get.call(this);
  yield this.regenerateSession();
  get.call(this);
}

app.listen(8080);
```

* After adding session middleware, you can use `this.session` to set or get the sessions.
* Getting session ID via `this.sessionId`.
* Setting `this.session = null;` will destroy this session.
* Altering `this.session.cookie` changes the cookie options of this user. Also you can use the cookie options in session the store. Use for example `cookie.maxAge` as the session store's ttl.
* Calling `this.regenerateSession` will destroy any existing session and generate a new, empty one in its place. The new session will have a different ID.
* Calling `this.saveSession` will save an existing session (this method was added for [koa-redirect-loop](https://github.com/niftylettuce/koa-redirect-loop))
* Setting `this.sessionSave = true` will force saving the session regardless of any other options or conditions.
* Setting `this.sessionSave = false` will prevent saving the session regardless of any other options or conditions.

### Options

 * `key`: cookie name defaulting to `koa.sid`.
 * `prefix`: session prefix for store, defaulting to `koa:sess:`.
 * `ttl`: ttl is for sessionStore's expiration time. it is different with `cookie.maxAge`, default to null(means get ttl from `cookie.maxAge`).
 * `rolling`: rolling session, always reset the cookie and sessions, defaults to `false`.
 * `genSid`: default sid was generated by [uid2](https://github.com/coreh/uid2), you can pass a function to replace it (supports promises/async functions).
 * `defer`: defers get session, only generate a session when you use it through `var session = yield this.session;`, defaults to `false`.
 * `allowEmpty`: allow generation of empty sessions.
 * `errorHandler(err, type, ctx)`: `Store.get` and `Store.set` will throw in some situation, use `errorHandle` to handle these errors by yourself. Default will throw.
 * `reconnectTimeout`: When store is disconnected, don't throw `store unavailable` error immediately, wait `reconnectTimeout` to reconnect, default is `10s`.
 * `sessionIdStore`: object with get, set, reset methods for passing session id throw requests.
 * `valid`: valid(ctx, session), valid session value before use it.
 * `beforeSave`: beforeSave(ctx, session), hook before save session.
 * `store`: session store instance. It can be any Object that has the methods `set`, `get`, `destroy` like [MemoryStore](https://github.com/koajs/koa-session/blob/master/lib/store.js).
 * `cookie`: session cookie settings, defaulting to:
    ```js
    {
      path: '/',
      httpOnly: true,
      maxAge: 24 * 60 * 60 * 1000 //one day in ms,
      overwrite: true,
      signed: true
    }
    ```

    For a full list of cookie options see [expressjs/cookies](https://github.com/expressjs/cookies#cookiesset-name--value---options--).

    if you set`cookie.maxAge` to `null`, meaning no "expires" parameter is set so the cookie becomes a browser-session cookie. When the user closes the browser the cookie (and session) will be removed.

    Notice that `ttl` is different from `cookie.maxAge`, `ttl` set the expire time of sessionStore. So if you set `cookie.maxAge = null`, and `ttl=ms('1d')`, the session will expired after one day, but the cookie will destroy when the user closes the browser.
    And mostly you can just ignore `options.ttl`, `koa-generic-session` will parse `cookie.maxAge` as the tll.

## Hooks

- `valid()`: valid session value before use it
- `beforeSave()`: hook before save session

## Session Store

You can use any other store to replace the default MemoryStore, it just needs to follow this api:

* `get(sid)`: get session object by sid
* `set(sid, sess, ttl)`: set session object for sid, with a ttl (in ms)
* `destroy(sid)`: destroy session for sid

the api needs to return a Promise, Thunk or generator.

And use these events to report the store's status.

* `connect`
* `disconnect`

### Stores Presented

- [koa-redis](https://github.com/koajs/koa-redis) to store your session data with redis.
- [koa-mysql-session](https://github.com/tb01923/koa-mysql-session) to store your session data with MySQL.
- [koa-generic-session-mongo](https://github.com/freakycue/koa-generic-session-mongo) to store your session data with MongoDB.
- [koa-pg-session](https://github.com/TMiguelT/koa-pg-session) to store your session data with PostgreSQL.
- [koa-generic-session-rethinkdb](https://github.com/KualiCo/koa-generic-session-rethinkdb) to store your session data with ReThinkDB.
- [koa-sqlite3-session](https://github.com/chichou/koa-sqlite3-session) to store your session data with SQLite3.
- [koa-generic-session-sequelize](https://github.com/natesilva/koa-generic-session-sequelize) to store your session data with the [Sequelize](http://docs.sequelizejs.com/) ORM.
- [koa-generic-session-knex](https://github.com/EarthlingInteractive/koa-generic-session-knex) to store your session data with the [Knex](http://knexjs.org/) query builder.

## Graceful shutdown

Since this middleware comes with an auto-reconnect feature, it's very likely you can't gracefully shutdown after
closing the client as generic-session will try to recover the connection, in those cases you can disable reconnect feature (https://github.com/koajs/generic-session/blob/49b45612877d1a1b8a42dc61bfeba46b71f9cb52/src/session.js#L103-L112) desactivating the client emitter (do this only when stopping the server)

Example with ioredis
```js
  // ...disconnecting from db
  redisClient.emit = () => true;
  await redisClient.quit();
  // ...stopping the server
```

## Licences
(The MIT License)

Copyright (c) 2013 - 2016 dead-horse and other contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
**This has moved to Releases, please see <https://github.com/koajs/generic-session/releases>.**

2.0.0 / 2017-07-11
==================

* Refactor to koa 2 - Supports node 6.x and 7.x (#123)

1.11.6 / 2017-06-12
==================

* fix: cookie options may be changed

1.11.5 / 2017-01-16
==================

  * fix: store use options.maxAge (#112)

1.11.4 / 2016-12-01
==================

  * docs: maxage => maxAge
  * fix: maxage => maxAge (#107)
  * update default cookie maxage in readme (#84)
  * add koa-generic-session-sequelize session store (#103)

1.11.3 / 2016-07-24
==================

  * fix: always refresh session (#95)

1.11.2 / 2016-07-12
==================

  * fix: npm pack should not include test dir (#92)

1.11.1 / 2016-06-23
==================

  * fix: make sure ctx.session exists when cookie.path = '/' (#91)

1.11.0 / 2016-06-20
==================

  * Add support for overriding session save conditions per request (#89)

1.10.2 / 2016-03-23
==================

  * fix: check if session exits

1.10.1 / 2015-12-15
==================

  * feat: ctx.session as a setter/getter

1.10.0 / 2015-11-15
==================

  * feat: support valid/beforeSave hooks

1.9.2 / 2015-08-26
==================

  * fix: support custom cookie options

1.9.1 / 2015-08-23
==================

  * Fix typo of 'available'

1.9.0 / 2015-06-05
==================

  * Added session id storage option to override session id store logic.

1.8.0 / 2015-03-15
==================

  * feat: support reconnectTimeout

1.7.0 / 2015-03-01
==================

  * Add support for regenerating sessions.

1.6.0 / 2015-01-16
==================

  * Merge pull request #39 from rfink/sess_id_optional_check
  * Allow a forced session id that is not available in the cookie

1.5.0 / 2015-01-07
==================

  * Merge pull request #37 from rfink/session_id_context
  * Bind genSid to context, add tests

1.4.0 / 2015-01-06
==================

  * fix(sessionId): ensure session id generate before yield next
  * fix indent

1.3.0 / 2014-11-22
==================

  * support options.errorHanlder, default throw set error

1.2.3 / 2014-11-17
==================

  * Delete session-id cookie if session not loaded.
  * export the session store object

1.2.2 / 2014-09-16
==================

  * use crc instead of buffer-crc32

1.2.1 / 2014-08-11
==================

  * use parseurl get the original url pathname

1.2.0 / 2014-08-09
==================

  * Merge pull request #25 from mekwall/patch-2
  * Added description for allowEmpty option
  * Added option to allow generation of empty session

1.1.3 / 2014-08-05
==================

  * Merge pull request #23 from mekwall/patch-1
  * Fixes MemoryStore not being exported

1.1.2 / 2014-07-07
==================

  * delete cookie.maxAge

1.1.1 / 2014-07-07
==================

  * fix maxage compat

1.1.0 / 2014-07-07
==================

  * seperate ttl and cookie.maxage

1.0.0 / 2014-06-29
==================

  * refactor
  * rename to koa-generic-session

0.4.1 / 2014-06-29
==================

  * warn rename to koa-generic-session

0.4.0 / 2014-06-18
==================

  * use uid-safe to generate default sid
  * maxAge -> maxage, close #18
  * Merge pull request #17 from yoshuawuyts/patch-1
  * Update README.md

0.3.3 / 2014-05-23
==================

  * bump dependencies

0.3.2 / 2014-05-22
==================

  * support options.genSid, fixed #16

0.3.1 / 2014-03-27
==================

  * fix store options bug, fixed #13

0.3.0 / 2014-03-15
==================

  * Merge pull request #11 from dead-horse/issue9-rolling-session
  * fix readme
  * add rolling session options, fixed #9
  * Merge pull request #10 from dead-horse/issue8-refresh
  * fixed #8, only gen cookie and set session when session is really modified

0.2.1 / 2014-03-14
==================

  * fix path error, fix defer session setter

0.2.0 / 2014-03-14
==================

  * Merge pull request #7 from dead-horse/issue6-defer
  * remove coverage
  * update readme
  * fix travis
  * finish defer, fixed #6
  * get session return status
  * add test for defer
  * add defer session

0.1.0 / 2014-02-27
==================

  * fix cookie.expires

0.1.0-beta1 / 2014-02-27
==================

  * update readme
  * seperate Store

0.0.9 / 2014-02-27
==================

  * fix new session setCookie twice

0.0.8 / 2014-02-11
==================

  * Merge pull request #3 from dead-horse/fix-cookie
  * get session error, throw it
  * only set the cookie when sessin was modified
  * Changed position of cookies.set
  * fix test
  * fix typo

0.0.7 / 2014-02-01
==================

  * fix maxage and compat connect type maxAge
  * add autod

0.0.6 / 2014-01-20
==================

  * when store is disconnect, throw 500

0.0.5 / 2013-12-30
==================

  * add middware name session

0.0.4 / 2013-12-29
==================

  * fix package.json

0.0.3 / 2013-12-29
==================

  * update readme
  * add cookie in session

0.0.2 / 2013-12-28
==================

  * Release 0.0.2
  * update readme
  * remove secret
  * use default cookie's signature

0.0.1 / 2013-12-28
==================

  * update test
  * add travis yml
  * update test
  * init session with memory store
  * Initial commit


src/session.js
/**!
 * koa-generic-session - lib/session.js
 * Copyright(c) 2013 - 2014
 * MIT Licensed
 *
 * Authors:
 *   dead_horse <dead_horse@qq.com> (http://deadhorse.me)
 */



/**
 * Module dependencies.
 */

const debug = require('debug')('koa-generic-session:session')
const MemoryStore = require('./memory_store')
const crc32 = require('crc').crc32
const parse = require('parseurl')
const Store = require('./store')
const copy = require('copy-to')
const uid = require('uid-safe')

/**
 * Warning message for `MemoryStore` usage in production.
 */

const warning = 'Warning: koa-generic-session\'s MemoryStore is not\n' +
  'designed for a production environment, as it will leak\n' +
  'memory, and will not scale past a single process.'

const defaultCookie = {
  httpOnly: true,
  path: '/',
  overwrite: true,
  signed: true,
  maxAge: 24 * 60 * 60 * 1000 //one day in ms
}

/**
 * setup session store with the given `options`
 * @param {Object} options
 *   - [`key`] cookie name, defaulting to `koa.sid`
 *   - [`store`] session store instance, default to MemoryStore
 *   - [`ttl`] store ttl in `ms`, default to oneday
 *   - [`prefix`] session prefix for store, defaulting to `koa:sess:`
 *   - [`cookie`] session cookie settings, defaulting to
 *     {path: '/', httpOnly: true, maxAge: null, overwrite: true, signed: true}
 *   - [`defer`] defer get session,
 *   - [`rolling`]  rolling session, always reset the cookie and sessions, default is false
 *     you should `await ctx.session` to get the session if defer is true, default is false
 *   - [`genSid`] you can use your own generator for sid (supports promises/async functions)
 *   - [`errorHandler`] handler for session store get or set error
 *   - [`valid`] valid(ctx, session), valid session value before use it
 *   - [`beforeSave`] beforeSave(ctx, session), hook before save session
 *   - [`sessionIdStore`] object with get, set, reset methods for passing session id throw requests.
 */

module.exports = function(options = {}) {

  const key = options.key || 'koa.sid'
  const client = options.store || new MemoryStore()
  const errorHandler = options.errorHandler || defaultErrorHanlder
  const reconnectTimeout = options.reconnectTimeout || 10000

  const store = new Store(client, {
    ttl: options.ttl,
    prefix: options.prefix
  })

  const genSid = options.genSid || uid;
  const valid = options.valid || noop
  const beforeSave = options.beforeSave || noop

  const cookie = options.cookie || {}
  copy(defaultCookie).to(cookie)

  let storeStatus = 'available'
  let waitStore = Promise.resolve()

  // notify user that this store is not
  // meant for a production environment
  if ('production' === process.env.NODE_ENV && client instanceof MemoryStore) {
    // eslint-disable-next-line
    console.warn(warning)
  }

  const sessionIdStore = options.sessionIdStore || {

    get: function() {
      return this.cookies.get(key, cookie)
    },

    set: function(sid, session) {
      this.cookies.set(key, sid, session.cookie)
    },

    reset: function() {
      this.cookies.set(key, null, { expires: new Date(0) })
    }
  }

  store.on('disconnect', () => {
    if (storeStatus !== 'available') return
    storeStatus = 'pending'
    waitStore = new Promise((resolve, reject) => {
      setTimeout(() => {
        if (storeStatus === 'pending') storeStatus = 'unavailable'
        reject(new Error('session store is unavailable'))
      }, reconnectTimeout)
      store.once('connect', resolve)
    })

  })

  store.on('connect', () => {
    storeStatus = 'available'
    waitStore = Promise.resolve()
  })

  // save empty session hash for compare
  const EMPTY_SESSION_HASH = hash(generateSession())

  return options.defer ? deferSession : session

  function addCommonAPI(ctx) {

    ctx._sessionSave = null

    // more flexible
    Object.defineProperty(ctx, 'sessionSave', {
      get: () => {
        return ctx._sessionSave
      },
      set: (save) => {
        ctx._sessionSave = save
      }
    })

  }

  /**
   * generate a new session
   */
  function generateSession() {
    const session = {}
    //you can alter the cookie options in nexts
    session.cookie = {}
    for (const prop in cookie) {
      session.cookie[prop] = cookie[prop]
    }
    compatMaxage(session.cookie)
    return session
  }

  /**
   * check url match cookie's path
   */
  function matchPath(ctx) {
    const pathname = parse(ctx).pathname
    const cookiePath = cookie.path || '/'
    if (cookiePath === '/') {
      return true
    }
    if (pathname.indexOf(cookiePath) !== 0) {
      debug('cookie path not match')
      return false
    }
    return true
  }

  /**
   * get session from store
   *   get sessionId from cookie
   *   save sessionId into context
   *   get session from store
   */
  async function getSession(ctx) {
    if (!matchPath(ctx)) return
    if (storeStatus === 'pending') {
      debug('store is disconnect and pending')
      await waitStore
    } else if (storeStatus === 'unavailable') {
      debug('store is unavailable')
      throw new Error('session store is unavailable')
    }

    if (!ctx.sessionId) {
      ctx.sessionId = sessionIdStore.get.call(ctx)
    }

    let session
    let isNew = false
    if (!ctx.sessionId) {
      debug('session id not exist, generate a new one')
      session = generateSession()
      ctx.sessionId = await genSid.call(ctx, 24)
      isNew = true
    } else {
      try {
        session = await store.get(ctx.sessionId)
        debug('get session %j with key %s', session, ctx.sessionId)
      } catch (err) {
        if (err.code === 'ENOENT') {
          debug('get session error, code = ENOENT')
        } else {
          debug('get session error: ', err && err.message)
          errorHandler(err, 'get', ctx)
        }
      }
    }

    // make sure the session is still valid
    if (!session ||
      !valid(ctx, session)) {
      debug('session is empty or invalid')
      session = generateSession()
      ctx.sessionId = await genSid.call(ctx, 24)
      sessionIdStore.reset.call(ctx)
      isNew = true
    }

    // get the originHash
    const originalHash = !isNew && hash(session)

    return {
      originalHash: originalHash,
      session: session,
      isNew: isNew
    }
  }

  /**
   * after everything done, refresh the session
   *   if session === null; delete it from store
   *   if session is modified, update cookie and store
   */
  async function refreshSession (ctx, session, originalHash, isNew) {

    // reject any session changes, and do not update session expiry
    if(ctx._sessionSave === false) {
      return debug('session save disabled')
    }

    //delete session
    if (!session) {
      if (!isNew) {
        debug('session set to null, destroy session: %s', ctx.sessionId)
        sessionIdStore.reset.call(ctx)
        return store.destroy(ctx.sessionId)
      }
      return debug('a new session and set to null, ignore destroy')
    }

    // force saving non-empty session
    if(ctx._sessionSave === true) {
      debug('session save forced')
      return saveNow(ctx, ctx.sessionId, session)
    }

    const newHash = hash(session)
    // if new session and not modified, just ignore
    if (!options.allowEmpty && isNew && newHash === EMPTY_SESSION_HASH) {
      return debug('new session and do not modified')
    }

    // rolling session will always reset cookie and session
    if (!options.rolling && newHash === originalHash) {
      return debug('session not modified')
    }

    debug('session modified')

    await saveNow(ctx, ctx.sessionId, session)

  }

  async function saveNow(ctx, id, session) {
    compatMaxage(session.cookie)

    // custom before save hook
    beforeSave(ctx, session)

    //update session
    try {
      await store.set(id, session)
      sessionIdStore.set.call(ctx, id, session)
      debug('saved')
    } catch (err) {
      debug('set session error: ', err && err.message)
      errorHandler(err, 'set', ctx)
    }
  }

  /**
   * common session middleware
   * each request will generate a new session
   *
   * ```
   * let session = this.session
   * ```
   */
  async function session(ctx, next) {
    ctx.sessionStore = store
    if (ctx.session || ctx._session) {
      return next()
    }
    const result = await getSession(ctx)
    if (!result) {
      return next()
    }

    addCommonAPI(ctx)

    ctx._session = result.session

    // more flexible
    Object.defineProperty(ctx, 'session', {
      get() {
        return this._session
      },
      set(sess) {
        this._session = sess
      }
    })

    ctx.saveSession = async function saveSession() {
      const result = await getSession(ctx)
      if (!result) {
        return next()
      }
      return refreshSession(ctx, ctx.session, result.originalHash, result.isNew)
    }

    ctx.regenerateSession = async function regenerateSession() {
      debug('regenerating session')
      if (!result.isNew) {
        // destroy the old session
        debug('destroying previous session')
        await store.destroy(ctx.sessionId)
      }

      ctx.session = generateSession()
      ctx.sessionId = await genSid.call(ctx, 24)
      sessionIdStore.reset.call(ctx)

      debug('created new session: %s', ctx.sessionId)
      result.isNew = true
    }

    // make sure `refreshSession` always called
    let firstError = null
    try {
      await next()
    } catch (err) {
      debug('next logic error: %s', err && err.message)
      firstError = err
    }
    // can't use finally because `refreshSession` is async
    try {
      await refreshSession(ctx, ctx.session, result.originalHash, result.isNew)
    } catch (err) {
      debug('refresh session error: %s', err && err.message)
      if (firstError) ctx.app.emit('error', err, ctx)
      firstError = firstError || err
    }
    if (firstError) throw firstError
  }

  /**
   * defer session middleware
   * only generate and get session when request use session
   *
   * ```
   * let session = yield this.session
   * ```
   */
  async function deferSession(ctx, next) {
    ctx.sessionStore = store

    // TODO:
    // Accessing ctx.session when it's defined is causing problems
    // because it has side effect. So, here we use a flag to determine
    // that session property is already defined.
    if (ctx.__isSessionDefined) {
      return next()
    }
    let isNew = false
    let originalHash = null
    let touchSession = false
    let getter = false

    // if path not match
    if (!matchPath(ctx)) {
      return next()
    }

    addCommonAPI(ctx)

    Object.defineProperty(ctx, 'session', {
      async get() {
        if (touchSession) {
          return this._session
        }
        touchSession = true
        getter = true

        const result = await getSession(this)
        // if cookie path not match
        // this route's controller should never use session
        if (!result) return

        originalHash = result.originalHash
        isNew = result.isNew
        this._session = result.session
        return this._session
      },
      set(value) {
        touchSession = true
        this._session = value
      }
    })

    // internal flag to determine that session is already defined
    ctx.__isSessionDefined = true

    ctx.saveSession = async function saveSession() {
      // make sure that the session has been loaded
      await ctx.session
      
      const result = await getSession(ctx)
      if (!result) {
        return next()
      }
      return refreshSession(ctx, ctx.session, result.originalHash, result.isNew)
    }

    ctx.regenerateSession = async function regenerateSession() {
      debug('regenerating session')
      // make sure that the session has been loaded
      await ctx.session

      if (!isNew) {
        // destroy the old session
        debug('destroying previous session')
        await store.destroy(ctx.sessionId)
      }

      ctx._session = generateSession()
      ctx.sessionId = await genSid.call(ctx, 24)
      sessionIdStore.reset.call(ctx)
      debug('created new session: %s', ctx.sessionId)
      isNew = true
      return ctx._session
    }

    await next()

    if (touchSession) {
      // if only this.session=, need try to decode and get the sessionID
      if (!getter) {
        ctx.sessionId = sessionIdStore.get.call(ctx)
      }

      await refreshSession(ctx, ctx._session, originalHash, isNew)
    }
  }
}

/**
 * get the hash of a session include cookie options.
 */
function hash(sess) {
  return crc32.signed(JSON.stringify(sess))
}

/**
 * cookie use maxAge, hack to compat connect type `maxage`
 */
function compatMaxage(opts) {
  if (opts) {
    opts.maxAge = opts.maxage ? opts.maxage : opts.maxAge
    delete opts.maxage
  }
}

module.exports.MemoryStore = MemoryStore

function defaultErrorHanlder (err, type) {
  err.name = 'koa-generic-session ' + type + ' error'
  throw err
}

function noop () {
  return true
}
